### ISSUE TITLE
Porting to Wii U/3DS

### ISSUE THREAD

**Issue by SeleDreams** [Rinnegatamante/vitaGL#85]
How complex do you think porting it to the Wii U and/or 3DS homebew SDKs would be ?
Because both kind of lack real opengl support apart from opengl 1.1 which makes porting game engines etc to the platforms fairly difficult

**Comment by Rinnegatamante** [comment]
It's not something you can port since it uses sceGxm which is proprietary Vita graphics API. It can be used however as implementation reference for similar projects.
3DS btw has picaGL made by MasterFeizz. ( https://github.com/masterfeizz/picaGL ).
Consider that 3DS has no programmable GPU (if i remember correctly, only vertex stage is programmable) which means a GL2/GLES2 implementation is not possible at all.

**Comment by SeleDreams** [comment]
> It's not something you can port since it uses sceGxm which is proprietary Vita graphics API. It can be used however as implementation reference for similar projects. 3DS btw has picaGL made by MasterFeizz. ( https://github.com/masterfeizz/picaGL ). Consider that 3DS has no programmable GPU (if i remember correctly, only vertex stage is programmable) which means a GL2/GLES2 implementation is not possible at all.

when I said port I meant it like by using the source but replacing all sceGxm calls by calls to the GX2 api for the wii u (or citro3d for the 3ds but that seems not possible)

**Comment by Rinnegatamante** [comment]
It's a question more suited for a Wii U developer. I've no idea what GX2 is and i'm not even remotely interested in developing something for a Nintendo console.

**Comment by SeleDreams** [comment]
> It's a question more suited for a Wii U developer. I've no idea what GX2 is and i'm not even remotely interested in developing something for a Nintendo console.

Just to respond tho, GX2 is basically the graphics processor of the console

**Comment by Rinnegatamante** [comment]
> > It's not something you can port since it uses sceGxm which is proprietary Vita graphics API. It can be used however as implementation reference for similar projects. 3DS btw has picaGL made by MasterFeizz. ( https://github.com/masterfeizz/picaGL ). Consider that 3DS has no programmable GPU (if i remember correctly, only vertex stage is programmable) which means a GL2/GLES2 implementation is not possible at all.
> 
> when I said port I meant it like by using the source but replacing all sceGxm calls by calls to the GX2 api for the wii u (or citro3d for the 3ds but that seems not possible)

You can't just replace sceGxm calls blindly. sceGxm has several differences compared to openGL that had been accounted for in vitaGL that other graphics api may not have. For example:
- Viewport sceGxm algorithm is different from openGL one: https://github.com/Rinnegatamante/vitaGL/blob/master/source/misc.c#L288-L296
- sceGxm has a special fragment program that allows to have a special stencil mode which consists in only a single bit related to on/off state. Such fragment program (called Mask Update program) can run together with normal stencil test and the related calculation is performed prior rasterization. This feature is used by vitaGL to implement efficient and fast pixel granularity scissor testing: https://github.com/Rinnegatamante/vitaGL/blob/master/source/tests.c#L250-L310
- Blending on sceGxm requires runtime patching of the shaders, so whenever a blend mode state changes, shaders need to be patched with sceGxmShaderPatcher to account for the change: https://github.com/Rinnegatamante/vitaGL/blob/master/source/custom_shaders.c#L250-L255

Those are just some examples of code that likely requires to be completely rewrote or accounted in a different way with other graphics API.
In short: vitaGL should really be used only as an implementation reference and not blindly attempt to replace sceGxm calls with another graphics API (Unless who's working on such "port" is also an expert of sceGxm).
